File: trunk/ARTServer/src/graphics/Scene.java
package graphics;

import graphics.drawables.IDrawable;
import graphics.drawables.ImageDrawable;

import java.util.Hashtable;
import java.util.Iterator;
import java.util.LinkedList;


public final class Scene
{
	public static final String LAYER_NAME_BACKGROUND	= "b";
	public static final String LAYER_NAME_FOREGROUND	= "f";
	public static final String LAYER_NAME_OVERLAY		= "o";
	public static final String LAYER_NAME_WIDGETS		= "w";
	
	private boolean _isDrawing = false;
	private final double _width;
	public double getWidth() { return _width; }
	
	private final double _height;
	public double getHeight() { return _height; }

	private final double _scale;
	public final SceneContext params;
	
	public final class Layer
	{
		public IDrawable clipMask = null;
		private LinkedList<IDrawable> _drawables = new LinkedList<IDrawable>();
		public boolean add(IDrawable drawable)
		{
			_drawables.add(drawable);
			return true;
		}
		
		public boolean remove(IDrawable drawable)
		{
			return _drawables.remove(drawable);
		}
		
		public Iterator<IDrawable> getDrawableIterator() {
			return _drawables.iterator();
		}
		
		void onDraw(SceneContext params)
		{
	        params.context.save();
	        if(this.clipMask != null) this.clipMask.onDraw(params);

	        for(IDrawable drawable : _drawables)
	        {
	        	if(drawable instanceof ImageDrawable)
	        	{
	        		drawable.onDraw(params);
	        	}
	        }

	        for(IDrawable drawable : _drawables)
	        {
	        	if(!(drawable instanceof ImageDrawable))
	        	{
	        		drawable.onDraw(params);
	        	}
	        }
	        
	        params.context.restore();
		}
	}
	
	private Hashtable<String, Layer> _layers = new Hashtable<String, Layer>();
	private LinkedList<Layer> _zorder = new LinkedList<Layer>();
	
	public Layer addLayer(String name)
	{
        Layer newLayer = new Layer(); 
        _zorder.addLast(newLayer);
        _layers.put(name, newLayer);
        return  newLayer;
	}
	
	public Layer getLayer(String name)
	{
		return _layers.get(name);
	}
	
	public void draw()
	{
		if(_isDrawing) return;

		_isDrawing = true;
        params.context.save();

		try
    	{	        
			params.context.translate(params.context.getWidth() / 2.0, params.context.getHeight() / 2.0);
					
			params.context.scale(_scale, _scale);
			
	        for(Layer layer : _zorder)
	        {
	        	layer.onDraw(params);
	        }
    	}
    	finally
    	{
        	params.context.restore();
        	_isDrawing = false;
    	}
	}
	
	public Scene(double width, double height, double scale, JavaCanvas canvas)
	{
		_width = width;
		_height = height;
		_scale = scale;
		double offsetX = _width / 2.0;
		double offsetY = _width / 2.0;
		params = new SceneContext(
				new Rect2D( -offsetX, -offsetY, offsetX, offsetY ), 
    			canvas
    		);
	}
}

File: tags/120921/js/design/scene.js
function RectangularClipMask(width, height)
{
    this.onDraw = function(params)
    {
        var hw = width / 2;
        var hh = height / 2;

        var ctx = params.context;
        
        ctx.beginPath();
		ctx.moveTo(-hw, -hh);
        ctx.lineTo(hw, -hh);
        ctx.lineTo(hw, hh);
        ctx.lineTo(-hw, hh);
		ctx.closePath();
        
        ctx.fillStyle = params.palette.paper;
        ctx.fill();
        ctx.clip();
    };
}

function CircularClipMask(width, height)
{
	var radius = Math.min(width, height) / 2;
    this.onDraw = function(params)
    {
        var ctx = params.context;
        ctx.beginPath();
        ctx.arc(0,0,radius,0,Math.PI*2,false);
        ctx.fillStyle = params.palette.paper;
        ctx.fill();
        ctx.clip();
    };
}

function SceneLayer(scene)
{
    var drawables = [];
    
    
    this.visible = true;
    this.clipMask = null;
    
    this.add = function(drawable)
    {
        drawables.push(drawable);
        return true;            
    };
    
    this.remove = function(drawable)
    {
        for(var i in drawables)
        {
            if(drawables[i] === drawable)
            {                
                if(drawables.onRelease)
                {
                	drawables.onRelease(scene.getCommonParams());
                }
                drawables.splice(i,1);
                return true;
            }           
        }
        return false;           
    };
    
    this.onDraw = function(params)
    {
        params.context.save();
        if(this.clipMask != null) this.clipMask.onDraw(params);

        for(var i=0; i<drawables.length; i++)
        {
            if(params.filter & drawables[i].displayGroup)
            {
            	drawables[i].onDraw(params);            	
            }
        }
                
        params.context.restore();
    };  

    this.onPress = function(params)
    {
        for(var i = drawables.length-1; i>=0; i--) 
        {
            if(drawables[i].onPress)
            {
                var capture = drawables[i].onPress(params); 
                if(capture) return capture;             
            }
        }
        return null;
    };

    this.onRelease = function(params)
    {
        for(var i in drawables)
        {
            if(drawables[i].onRelease)
            {
                var capture = drawables[i].onRelease(params); 
                if(capture) return capture;             
            }
        }
        return null;
    };

    this.onMove = function(params)
    {
        for(var i in drawables)
        {
            if(drawables[i].onMove)
            {
                var capture = drawables[i].onMove(params); 
                if(capture) return capture;             
            }
        }
        return null;
    };
}


function Scene(canvas, width, height)
{
    var layers = {};
    var zorder = [];
    var capture = null;
    var context = null;
    
	this.inkColor = "black";//"#000000";
    
    

    this.getCommonParams = function()
    {
        return {
            dirty : false,
            scale : this.scale,
            filter : Scene.DISPLAY_GROUP_ANY,
            palette : {
            	ink: this.inkColor,
            	paper: "#ffffff",
            	background: "#404040",
            	highlight:"rgba(180,240,255,0.75)"
            },
            viewRect : {
                x1 : -this.offsetX,
                y1 : -this.offsetY,
                x2 : -this.offsetX + this.width / this.scale,               
                y2 : -this.offsetY + this.height / this.scale               
            }
        };
    }

    this.canvas = canvas;
    this.scale = 1.0;
    this.offsetX = width / 2;
    this.offsetY = height / 2;
    this.width = width;
    this.height = height;
    this.backgroundColor = null;
    
    this.addLayer = function(name)
    {
        var newLayer = new SceneLayer(this); 
        zorder.push(newLayer);
        layers[name] = newLayer;
        return  newLayer;
    };
    
    this.setLayer = function(name, newLayer)
    {
		var oldLayer = layers[name]; 
		layers[name] = newLayer;
		for(var i in zorder)
		{
			if(zorder[i] === oldLayer)
			{
				zorder[i] = newLayer;
				break;
			}
		}
    };

    this.getLayer = function(name)
    {
        return layers[name];
    };
    
    this.onPress = function(x, y)
    {
        var params = this.getCommonParams();
        params.x = (x - this.offsetX) / this.scale;
        params.y = (y - this.offsetY) / this.scale;
        
        for(params.pass=0; params.pass<2; params.pass++)
        {
            //for(var i = 0; i < zorder.length; i++) 
            for(var i = zorder.length-1; i>= 0; i--) 
            {
                capture = zorder[i].onPress(params); 
                if(capture) break;              
            }
            if(capture != null) break;
        }

        if(params.dirty) this.redraw();
        return capture; 
    };

    this.onRelease = function(x, y)
    {
        var params = this.getCommonParams();
        params.x = (x - this.offsetX) / this.scale;
        params.y = (y - this.offsetY) / this.scale;
        
        if(capture)
        {
            capture.onRelease(params);
        }
        else
        {
            for(var i = zorder.length-1; i>= 0; i--) 
            {
                if(zorder[i].onRelease(params)) break; 
            }
        }
        capture = null;

        if(params.dirty) this.redraw(); 
    };

    this.onMove = function(x, y)
    {
        var params = this.getCommonParams();
        params.x = (x - this.offsetX) / this.scale;
        params.y = (y - this.offsetY) / this.scale;
        
        if(capture)
        {
            capture.onMove(params);
        }
        else
        {
            for(var i = zorder.length-1; i>= 0; i--) 
            {
                if(zorder[i].onMove(params)) break; 
            }
        }

        if(params.dirty) this.redraw(); 
        return capture; 
    };

    this.redraw = function(filter)
    {
        var canvas = this.canvas;        
        var params = this.getCommonParams();
        
		if(filter) params.filter = filter;
				        
        params.context = canvas.getContext("2d");
        
        if(this.backgroundColor)
        {
        	params.context.fillStyle = this.backgroundColor;
        	params.context.fillRect( 0 , 0 , canvas.width , canvas.height );        	
        }
        else
        {
            params.context.clearRect( 0 , 0 , canvas.width , canvas.height );        	
        }
        
        params.context.save();
		params.context.translate(this.offsetX, this.offsetY);
		params.context.scale(this.scale,this.scale);
		
        for(var i=0; i<zorder.length; i++)
        {
            if(zorder[i].visible)
            {
                zorder[i].onDraw(params);
            }
        }
        params.context.restore();
    };

    this.addLayer(Scene.LAYER_BACKGROUND);
    this.addLayer(Scene.LAYER_FOREGROUND);
    this.addLayer(Scene.LAYER_OVERLAY);
    this.addLayer(Scene.LAYER_WIDGETS);
}

Scene.LAYER_BACKGROUND  = "b";
Scene.LAYER_FOREGROUND  = "f";
Scene.LAYER_OVERLAY     = "o";
Scene.LAYER_WIDGETS     = "w";

Scene.DISPLAY_GROUP_CONTENT		= 1;
Scene.DISPLAY_GROUP_UI			= 2;
Scene.DISPLAY_GROUP_ANY			= 3;